### ISSUE TITLE
Unable to skip already signed files (for instance for publishing a website zip)

### ISSUE THREAD

**Issue by CZEMacLeod** [dotnet/sign#647]
It would be useful when signing e.g. a zip file created by publishing a website, to only sign unsigned files. This way existing signatures on DLLs such as those distributed by Microsoft don't get overwritten, and there would be less overhead when using KeyVault HSM.

This might be indirectly related to #547 where the default action was changed to overwrite, and the fact that the documented `force` flag doesn't seem to exist. The options to skip, overwrite, or append a signature would seem to be appropriate instead of force.

If you select to skip existing signatures, it should (optionally?) check the existing signature is valid, then skip the file if it is. If it is not, then it should 'fall through' to overwrite. In the case where append is used, it should also check the existing signature.

At the moment, the only way of implementing this that I can see, is to pre-scan the files and generate a filelist, although it is not entirely clear how this would work with files inside nuget or zip files as the documentation around the --file-list command is a little unclear.

Version: 0.9.1-beta.23530.1+f4efed9e8fb3296f29497b90feb6548e506f2078

**Comment by dtivel** [comment]
CC @clairernovotny

Authenticode supports multiple signatures.  It was planned that Sign CLI's default behavior was to append new signatures. 
 However, last we checked a Windows API we use (`SignerSignEx3(...)`) did not support appending; instead, it overwrote an existing signature.  The `--force` option in the draft specification is in a limbo state because we haven't a good solution yet.

That said, there is a way to skip already signed files using file lists.  You would use a file list to exclude already signed files in a separate invocation of Sign CLI.  File lists have one file pattern per line.  Each pattern can be a glob pattern.

**Comment by CZEMacLeod** [comment]
@dtivel Yes - I had worked out that the filelist might be a way of doing it, however, I run the command on a directory with multiple .nuget files (sometimes up to 100), and to work out the list would have to open and scan each file in the archive to check if it is already signed and add the ones that aren't to the list. 
I'm not even sure how it would work with the multiple nuget files. Would I need to include a pattern to sign the nuget package itself too?
If this is the only available mechanism, then perhaps a good example of how to do this would be an idea?

As the command is already opening and unzipping the files to sign them, it would be nice to be able to test the files to check if they already have a valid signature, and only sign those without the signature. I appreciate that append and/or overwrite were the original intention - however, adding the third option to skip would be useful.

I appreciate this would be an extra step in the logic, and might slow it down somewhat. However, it would be much easier than basically doing the same work twice.

You mention using SignerSignEx3 - is this used in the azure key vault code? I assumed the signing was actually done in the cloud in this case, as the key cannot be exported?

**Comment by bradwilson** [comment]
Does the file list skip specification include the ability to exclude files that are inside a NuGet package? Is the list in that case a relative path to the file inside the NuGet package?

**Comment by bradwilson** [comment]
It's easier to just list the two files I want signed rather than the 6 I don't. 😂

**Comment by ilCosmico** [comment]
In my case the assemblies inside the NuGet package were already signed before packing, so they were supposed to be included as-is. Using the `-rc False` option prevented Sign CLI from re-signing the internal DLLs and allowed the package to be produced without altering the original binaries.

Posting this here in case it helps others facing the same situation.

**Comment by XVII** [comment]
For example, **specific** files in the MSIX need to be signed, i.e. `app.exe`, then the `MSIX` itself (container) `app.msix` needs to be signed, but nothing else should have its signature clobbered. At the moment `app.msix` is fully unpacked, every file (signed and unsigned) are signed then the container is signed.

The MSIX build and packaging process is one step in my pipeline, so it would be tricky to do a second sign between those steps (after `app.exe` is created and before `app.msix` is packaged).

It's almost like there would need to be different filters for recursion?
We'd definitely need some documentation on how containers are handled in regards to filters.

A filter of the following doesn't work because it signs all files in the container of `app.msix`:
I'd need these three files signed: `app.exe`, `app.dll`, `app.msix`.

`.\sign code azure-key-vault "**/*.msix" "app.*"`

Simply skipping "files already signed" seems like the easier option.

###  Update 1
We're actually seeing this start to impact end user performance due to over-signing.

A quick test using `Get-MpPerformanceReport`, shows that Microsoft-authored DLLs (e.g. `Microsoft.Windows.SDK.NET.dll`) that were originally signed by Microsoft but re-signed by our certificate are hitting ~5s `TrustCheck` delays _per_ DLL on app launch via Defender real-time protection. I assume that Microsoft-signed binaries are fast-tracked through Defender, but once the signature is overwritten with a lesser-known publisher cert, it could trigger a full cloud reputation/CRL lookup?

###  Update 2
Turns out that `--file-list` [does only work for containers](https://github.com/dotnet/sign/issues/990) and should probably documented or changed to `--container-file-list` to make it a bit clearer.

###  Update 3
 It also turns out that `PublishReadyToRun` (and `PublishSingleFile`) will recompile the .NET assemblies with native code which also causes the signature to be cleared. In case that is confusing anyone :)

**Comment by XVII** [comment]
@dtivel , just wondering if you've had any more thoughts on this one? I.e. updating the doco or renaming the parameter?
